### PR TITLE
Added serialization from etcd error metric

### PR DESCRIPTION
**Pull request by baomingwang**
#### What type of PR is this?

/kind bug
/sig api-machinery

#### What this PR does / why we need it:
To revive previous PR https://github.com/kubernetes/kubernetes/pull/90612 for capturing data corruption issue https://github.com/kubernetes/kubernetes/issues/69579

Able to manually flip one bit and have targeting metric `storage_decode_errors` emitted by replace `protoEncodingPrefix ` from `[]byte{0x6b, 0x38, 0x73, 0x00}` to `[]byte{0x6b, 0x38, 0x73, 0x01}`.

```
$ kubectl get pods nginx
Error from server: illegal base64 data at input

$ kubectl get --raw /metrics | grep storage_decode_errors
# HELP apiserver_storage_decode_errors_total [ALPHA] Number of stored object decode errors split by object type
# TYPE apiserver_storage_decode_errors_total counter
apiserver_storage_decode_errors_total{resource="pods"} 1
```

kube-apiserver log with --v=4 output
```
I1208 07:56:17.397972 10 store.go:181] Decoding pods "/pods/default/nginx" failed: illegal base64 data at input byte 3
```

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
kube-apiserver: errors decoding objects in etcd are now recorded in an `apiserver_storage_decode_errors_total` counter metric
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by baomingwang** [comment]
/assign @deads2k @lavalamp 
cc @jingyih @micahhausler who was working on previous PR

**Comment by dashpole** [comment]
/assign @dgrisonnet 
for instrumentation review

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go
Comment (dgrisonnet) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:118]: The metric should be prefixed by `apiserver`

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go
Comment (dgrisonnet) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:122]: `type` in an error metric is quite confusing since it might imply that it stores the type of the error, `resource` would be better suited and more inline with the existing metrics.

Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:122]: will update. thanks

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:118]: Got it. Will rename to `apiserver_storage_decode_errors`

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:161]: This would need to go in GetList too, at the very least

**Comment by lavalamp** [comment]
I'm not super sure this makes sense, the requests will already be failing and the log will have details in it. What does putting this in a metric accomplish?

This should essentially *never* happen. I have seen plenty of etcd corruption and it never is within-object corruption like this would detect.

/assign @deads2k @logicalhan 
/hold

I'd like those two people to at least look at this before merging it.

**Comment by baomingwang** [comment]
Hi @lavalamp 
Thank you for reply. Yes, the log should've already capture the error, but it would be much easier if metric could be emitted so as to alarm maintainer and bring the cluster back timely

**Comment by lavalamp** [comment]
If we were to do something to address this, I'm not sure a metric is the thing to do. Maybe auto delete the offending object? I still want @logicalhan and @deads2k to weigh in.

**Comment by logicalhan** [comment]
> This should essentially _never_ happen. I have seen plenty of etcd corruption and it never is within-object corruption like this would detect.

I agree with this. This happening should be pretty catastrophic. 

> Maybe auto delete the offending object? 

That could also be weird, imagine if the offending object was a secret or some important config map or something. Deleting is probably better than randomly blowing up though. 

I'm in the camp that we should probably record this occurring in a metric, despite the fact that it should be extremely low frequency. @lavalamp isn't that what you suggested [here](https://github.com/kubernetes/kubernetes/issues/69579#issuecomment-597894693)?

**Comment by wojtek-t** [comment]
> I'm in the camp that we should probably record this occurring in a metric, despite the fact that it should be extremely low frequency.

+1

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:120]:
```suggestion
            Namespace:      "apiserver"
			Name:           "storage_decode_errors_total",
			Help:           "Number of stored object decode errors split by object type",
			StabilityLevel: compbasemetrics.ALPHA,
```

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:120]: It's preferable to leave the component that the metric is in, in the `Namespace` field and counters should be conventionally suffixed with `_total`.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:161]: Added. Thank you!

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go]: Updated! Thanks!

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (cenkalti) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:161]: I think the correct place to capture this metric is inside `appendListItem` function, right after `codec.Decode` call because `appendListItem` does some other work after decoding.

Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:161]: MSTM. Will relocate inside of `appendListItem`

**Review by logicalhan — COMMENTED** [review]
/lgtm
/approve

from an instrumentation perspective.

**Comment by baomingwang** [comment]
@logicalhan @deads2k 
Do you mind take another took plz after all new comments were addressed? Thanks!
/cc @dims

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:164]: We should have a unit test that verifies this does what you want (i.e. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/prometheus/slis/metrics_test.go)

Comment (dims) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:164]: good point @logicalhan

Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:164]: thank you. `metrics_test.go` added

**Comment by baomingwang** [comment]
hi @logicalhan . Gentle reminder if you can take another look. Thanks!
/cc @dims

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics_test.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics_test.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics_test.go]: Good catch! Updated

**Comment by dims** [comment]
/approve
/lgtm

let's wait for @liggitt before we remove hold

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:162]: can any of the content we're trying to decode leak into `err`? if so, I suspect we shouldn't be logging it (or we could log content of Secret objects, etc)

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:1021]: can any of the content we're trying to decode leak into `err`? if so, I suspect we shouldn't be logging it (or we could log content of Secret objects, etc)

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:752]: would it be simpler to record/log inside this if block instead of plumbing `groupResource` and `key` down into `appendListItem` just for metrics/logging?

**Comment by liggitt** [comment]
should we have a test that flips a bit in etcd, then calls storage accessor methods and ensures this metric increments?

also, we decode in paths other than get and list (on every write, in watch, etc). should this metric get incremented if decode errors on existing data are encountered in those methods?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:752]: Yes. That's my original plan. 
But I got another suggestion from @cenkalti that even though the non-nil error that  `appendListItem()` returned can only from `obj, _, err := codec.Decode(data, nil, newItemFunc())` at this moment, it would make hard for future developers who want to add more functionality into `appendListItem()` to make another non-nil error as return.
Does it make sense?
https://github.com/kubernetes/kubernetes/pull/114376#discussion_r1088124683

Comment (cenkalti) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:752]: Another option would be move `decode()` from `appendListItem` to here.

**Comment by baomingwang** [comment]
> should we have a test that flips a bit in etcd, then calls storage accessor methods and ensures this metric increments?

I did some local testing to have targeting metric storage_decode_errors emitted by replace protoEncodingPrefix  from []byte{0x6b, 0x38, 0x73, 0x00} to []byte{0x6b, 0x38, 0x73, 0x01}.
```
$ kubectl get pods nginx
Error from server: illegal base64 data at input

$ kubectl get --raw /metrics | grep storage_decode_errors
# HELP storage_decode_errors [ALPHA] Number of stored object decode errors split by object type.
# TYPE storage_decode_errors counter
storage_decode_errors{type="pods"} 1

I1208 07:56:17.397972 10 store.go:181] Decoding pods "/pods/default/nginx" failed: illegal base64 data at input byte 3
```

Is it enough?

> also, we decode in paths other than get and list (on every write, in watch, etc). should this metric get incremented if decode errors on existing data are encountered in those methods?

Really nice callout. You remind me that it would be better to capture decode error inside of `decode()` function, to capture every `codec.Decode` errors. Does it make sense to you?
https://github.com/kubernetes/kubernetes/pull/114376/commits/a16529ccbf1a2af093f6ccffc97697a9e79ee99f#diff-83db812e87640d3b07e697ed074e5e7a8f349429abf033fbb34e98dd8b82dac5R1003

@liggitt

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:162]: I'm not quite sure if the decode error would leak to `err`, but we've always been logging the err when we called `decode()`. like 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L220-L224
Since it's been there for a long time, I think the leak issue should not be a major concern. WDYT?

**Comment by baomingwang** [comment]
Comments addressed with updated commit. Can you pls take a look again?
Let me know if anything didn't look to you.
@liggitt @dims

**Comment by dims** [comment]
/lgtm

will wait for @liggitt to remove hold!

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:159]:
```suggestion
	return decode(s.codec, s.groupResourceString, preparedKey, s.versioner, data, out, kv.ModRevision)
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:220]:
```suggestion
		err = decode(s.codec, s.groupResourceString, preparedKey, s.versioner, data, out, putResp.Header.Revision)
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:473]:
```suggestion
				return decode(s.codec, s.groupResourceString, preparedKey, s.versioner, origState.data, destination, origState.rev)
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:518]:
```suggestion
		err = decode(s.codec, s.groupResourceString, preparedKey, s.versioner, data, destination, putResp.Header.Revision)
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:994]: document the resource and key parameters are only intended to aid metrics/logging if the decode fails, they must not affect decode behavior at all

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:1001]: I still have an open question about whether content in `data` can leak into the error and get logged

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: I still think we need to know what we're logging... can you bit flip a json object and a proto object and see what comes back in the resulting error?

**Review by liggitt — COMMENTED** [review]
plumbing data to decode and recording the metric / error there makes sense. If we're intending to log the actual path in etcd, we should use `preparedKey` in a bunch of places

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Updated

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Updated

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Updated

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Updated

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:994]: Sorry I'm not sure I understand it correctly.
Do you mean we should not add parameters to decode() function? If yes, how should we pass in `resource`/`key` to record metric and logging?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: By flipping a protobuf object, I got the logging like
```
I1208 07:56:17.397972 10 store.go:181] Decoding pods "/pods/default/nginx" failed: illegal base64 data at input byte 3
```
The logging of err looks good.
I'm still not quite sure if error would be polluted (at least not during my reproducing), but it makes sense to me that we could avoid potential issues by not logging err.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Yes I did some reproducing, and the logging looks like this
```
I1208 07:56:17.397972 10 store.go:181] Decoding pods "/pods/default/nginx" failed: illegal base64 data at input byte 3
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:994]: passing them in is fine, but the parameters should be named and/or documented to make it clear their only intended purpose is for logging

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:994]: I don't like passing these in here, it's extremely confusing, why not make the call sites call a separate function to handle the error?

Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:994]: Thank you @liggitt . Clear to me now.

@lavalamp Do you mean keep func `decode()` unchanged, and for each time we call decode(...), wrap it like
```
err = decode(...)
if err != nil {
    recordDecodeError(resource, preparedKey)
}

func recordDecodeError(resource string, key string) {
    metrics.RecordDecodeError(resource)
    klog.V(4).Infof("Decoding %s \"%s\" failed", resource, key)
}
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (baomingwang) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Also, have tried to call bad object multiple times, the metric could accumulate correctly each time
```
$ kubectl get pods nginx
Error from server: illegal base64 data at input

$ kubectl get --raw /metrics | grep storage_decode_errors
# HELP storage_decode_errors [ALPHA] Number of stored object decode errors split by object type.
# TYPE storage_decode_errors counter
storage_decode_errors{type="pods"} 1
```
```
$ kubectl get pods nginx
Error from server: illegal base64 data at input

$ kubectl get --raw /metrics | grep storage_decode_errors
# HELP storage_decode_errors [ALPHA] Number of stored object decode errors split by object type.
# TYPE storage_decode_errors counter
storage_decode_errors{type="pods"} 2
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:994]: Yes, like that. Note for listed items, that preparedKey is not going to be helpful at all for debugging.

**Pull request merged**
